syntax = "proto3";

package youtubedl;

service YoutubeDL {
    rpc Ping (PingRequest) returns (PingReply);
    rpc AddToQueue(DownloadItem) returns (DownloadItemResponse);
    rpc RemoveFromQueue(DownloadID) returns (DownloadRemoveResponse);
    rpc GetAllDownloads(AllDownloadsRequest) returns (AllDownloads);
}

message PingRequest {
    string message = 1;
}

message PingReply {
    string message = 1;
}

message DownloadItem {
    string url = 1;
}

message DownloadItemResponse {
    string id = 1;
}

message DownloadID {
    string id = 1;
}

message DownloadRemoveResponse {
    string message = 1;
}

message AllDownloadsRequest {
}

message AllDownloads {
    repeated Download downloads = 1;
}

message Download {
    string url = 1;
    string title = 2;
    message DownloadState {
        string status = 1;
        string filename = 2;
        string eta_str = 3;
        string percent_str = 4;
        string speed_str = 5;
        string total_bytes_str = 6;
    }
    DownloadState state = 3;
}